pkg/azure/actuator.go

```diff
@@ -81,41 +104,480 @@ func isAzureCredentials(providerSpec *runtime.RawExtension) (bool, error) {
+		logger,
+		string(cloudCredsSecret.Data[AzureClientID]),
+		string(cloudCredsSecret.Data[AzureClientSecret]),
+		string(cloudCredsSecret.Data[AzureTenantID]),
+		string(cloudCredsSecret.Data[AzureSubscriptionID]),
+	)
+	if err != nil {
+		return fmt.Errorf("unable to create azure cred minter: %v", err)
+	}
+
+	// Client secret can not be retrieved from Azure, it can be only re-generated.
+	// Thus, either use already existing (if it can be found) or generate new one.
+	clientSecret := ""
+	existingSecret := &corev1.Secret{}
+	if err := a.client.Get(ctx, client.ObjectKey{Namespace: cr.Spec.SecretRef.Namespace, Name: cr.Spec.SecretRef.Name}, existingSecret); err == nil {
+		if existingSecret.ResourceVersion == azureStatus.SecretLastResourceVersion {
+			clientSecret = string(existingSecret.Data[AzureClientSecret])
+		}
+	}
+
+	addApp, newClientSecret, err := azureCredentialsMinter.CreateOrUpdateAADApplication(ctx, spName, clientSecret == "")
+	if err != nil {
+		return err
+	}
+
+	if clientSecret == "" {
+		clientSecret = newClientSecret
+	}
+
+	if addApp.AppID == nil {
+		return fmt.Errorf("service principal %q application ID is empty", spName)
+	}
+
+	azureStatus.AppID = *addApp.AppID
+	err = a.updateProviderStatus(ctx, logger, cr, azureStatus)
+	if err != nil {
+		return err
+	}
+
+	servicePrincipal, err := azureCredentialsMinter.CreateOrUpdateServicePrincipal(ctx, *addApp.AppID)
+	if err != nil {
+		return err
+	}
+
+	if servicePrincipal.DisplayName == nil {
+		return fmt.Errorf("service principal %q display name is empty", spName)
+	}
+
+	azureStatus.ServicePrincipalName = *servicePrincipal.DisplayName
+
+	err = a.updateProviderStatus(ctx, logger, cr, azureStatus)
+	if err != nil {
+		return err
+	}
+
+	if err := azureCredentialsMinter.AssignResourceScopedRole(ctx, infraResourceGroups, *servicePrincipal.ObjectID, *servicePrincipal.DisplayName, targetRole); err != nil {
+		return err
+	}
+
+	newSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cr.Spec.SecretRef.Name,
+			Namespace: cr.Spec.SecretRef.Namespace,
+			Annotations: map[string]string{
+				minterv1.AnnotationCredentialsRequest: fmt.Sprintf("%s/%s", cr.Namespace, cr.Name),
+			},
+		},
+		Data: map[string][]byte{
+			AzureClientID:       []byte(*servicePrincipal.AppID),
+			AzureClientSecret:   []byte(clientSecret),
+			AzureRegion:         cloudCredsSecret.Data[AzureRegion],
+			AzureResourceGroup:  cloudCredsSecret.Data[AzureResourceGroup],
+			AzureResourcePrefix: cloudCredsSecret.Data[AzureResourcePrefix],
+			AzureSubscriptionID: cloudCredsSecret.Data[AzureSubscriptionID],
+			AzureTenantID:       cloudCredsSecret.Data[AzureTenantID],
+		},
+	}
+
+	if err := a.syncCredentialSecrets(ctx, cr, newSecret, logger); err != nil {
+		return fmt.Errorf("unable to sync credetial secret: %v", err)
+	}
+
+	updatedSecret := &corev1.Secret{}
+	if err := wait.PollImmediate(2*time.Second, 10*time.Second, func() (bool, error) {
```

Comment:
This could use a comment, it looks like it's being done just to record the current resource version of the secret?